src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,235 @@
+            {
+                return default;
+            }
+
+            if (declarator.Initializer.Value is null ||
+                declarator.Initializer.Value.Kind == OperationKind.Invalid)
+            {
+                return default;
+            }
+
+            var isValidSelection = await CodeRefactoringHelpers.RefactoringSelectionIsValidAsync(
+                document,
+                selection,
+                node: declarationSyntax,
+                holes: ImmutableArray.Create(declarator.Initializer.Value.Syntax),
+                cancellationToken).ConfigureAwait(false);
+
+            if (!isValidSelection)
+            {
+                return default;
+            }
+
+            var disposableType = semanticModel.Compilation?.GetSpecialType(SpecialType.System_IDisposable);
+            if (disposableType is null || !localType.InheritsFromOrEquals(disposableType, includeInterfaces: true))
+            {
+                return default;
+            }
+
+            return (declarationSyntax, declarator.Symbol);
+        }
+
+        private async Task<Document> IntroduceUsingStatementAsync(
+            Document document,
+            TextSpan span,
+            CancellationToken cancellationToken)
+        {
+            var (declarationStatement, localVariable) = await FindDisposableLocalDeclaration(document, span, cancellationToken).ConfigureAwait(false);
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            if (!(declarationStatement.Parent is TBlockSyntax parent))
+            {
+                throw new NotImplementedException();
+            }
+
+            var lastUsageStatement = FindSiblingStatementContainingLastUsage(
+                declarationStatement,
+                localVariable,
+                semanticModel,
+                document.GetLanguageService<ISyntaxFactsService>(),
+                cancellationToken);
+
+            var parentStatementsList = GetStatements(parent);
+
+            var declarationStatementIndex = parentStatementsList.IndexOf(declarationStatement);
+
+            var lastUsageStatementIndex = lastUsageStatement is null ? -1 :
+                parentStatementsList.IndexOf(lastUsageStatement, declarationStatementIndex + 1);
+
+            var statementsToSurround = GetStatementsToSurround(parentStatementsList, declarationStatementIndex, lastUsageStatementIndex);
+
+            var newParent =
+                WithStatements(parent, parentStatementsList
+                    .ReplaceRange(
+                        index: declarationStatementIndex,
+                        count: 1 + statementsToSurround.Count,
+                        newNodes: new[]
+                        {
+                            CreateUsingStatement(declarationStatement, statementsToSurround)
+                        }));
```

Comment:
some of this nesting is hard to read.  tehre are both vertical and horizontal matching that has to happen.  it might be nice to either do things like align 'parent' and 'parentStatementList', or extract the second value into its own local to make it clearer what is going on.